### DISCUSSION TITLE
HBIOS Size Reduction

### DISCUSSION THREAD

**Discussion by kiwisincebirth** [wwarthen/RomWBW#478] [Ideas]
Just an observation about possible reduction of HBIOS size. It comes down to the routine PRTSUM which prints the device table including device attributes (which is 1372 bytes) according to the build log.

PRTSUM: is CALL'd from 2 places in HBIOS
* INITSYS3A: - which is after most initialisation (all devices) has completed, before it transitions to RomLdr
* a JP Vector in HB_ENTRYTBL at $400 - which I can only find use of in RomLDR in the code - `D` menu option

```
2231   8929             prtall:
2232   8929 3E 80       	ld	a,BID_BIOS		; BIOS Bank please
2233   892B DD 21 06 04 	ld	ix,$0406		; HBIOS PRTSUM vector
2234   892F C3 F9 FF    	jp	HB_BNKCALL		; do it
```

Proposal:
* Move the PRTSUM code from HBIOS to ROMLDR, replacing the above code
* Remove CALL to PRTSUM from HBIOS (INITSYS3A), and move it to the start of ROMLDR

Thus 
* When RomLdr start it prints the device table - user shouldn't notice any change
* RomLdr can still print it on-demand via `D` command, it is no longer a bank call.
* We save 1372 bytes in HBIOS, which i don't think is insignificant.

The only slight issue is that code (being moved) relies on hbios/util.asm which romldr needs to include

**Comment by wwarthen** [comment]
Hi @kiwisincebirth,

I completely agree with moving PRTSUM out of HBIOS proper.  Indeed, RomLdr is an appropriate place for it.  The problem is the issue with no space being available in RomLdr (see #477).

I am going to think about the space issue in RomLdr a bit.  However, I am not sure I have a good solution.

Thanks, Wayne

**Reply by kiwisincebirth** [reply]
Its actually not as good as I stated.

* You cannot delete everything in the PRTSUM segment, a couple of routines are utilities used in other code, the residual is 149 bytes left in hbios, (actual reduction is 1220 bytes) - still an improvement.
* copying the code into Romldr (while works for me, as i don't have the extra features enabled), it bring with it several include files most notably util.asm (not small), it would be abit of work to go through and individually copy only what is required, and do any consolidation.

FYI LDR_SIZ went from $1000 to $1B00 - with both ( incl PRTSUM ) sets of changes, and without any optimisation !!!

**Comment by wwarthen** [comment]
@kiwisincebirth,

Like my suggest for the slice inventory, this could be implemented in a new ROM app.

The only disadvantage of doing this in a ROM app is that the functionality would not be available in an Application Boot, but I don't see that as much of an issue.

Thanks, Wayne

**Reply by kiwisincebirth** [reply]
- Also the device summary wouldn't be printed (unless take steps) when the system first boots, today it it is displayed at end of hbios start, are you proposing to change this?

Ie. user no longer sees it unless they press D option from menu?

I am sure there are reasons why Romldr is compiled into the app boot file and not loaded from rom, assume it is by design because you want to be able to test the end to end boot process before flashing it to rom, makes sense of course 

I am also had thoughts about a single rom app for both functions D and S with just a different entry point. 

Another thought would be to load this app (once) into mid memory, but not conflict with RomLdr itself , and leave it resident, so could directly be called. 

Sort of RomLdr companion app, with non critical (to boot) functions. 

For app boot this would have to be stubbed out

Note I am physically away after Christmas, until early in new year, so wont be able to action this until then.

Mark

**Reply by wwarthen** [reply]
> * Also the device summary wouldn't be printed (unless take steps) when the system first boots, today it it is displayed at end of hbios start, are you proposing to change this?
> 
> Ie. user no longer sees it unless they press D option from menu?

The device summary can still be printed when RomLdr starts.  It is trivial to exec a ROM App and have control come back to RomLdr.  It doesn't need to be based on a keypress.

> I am sure there are reasons why Romldr is compiled into the app boot file and not loaded from rom, assume it is by design because you want to be able to test the end to end boot process before flashing it to rom, makes sense of course

Correct.  The primary intent is to be able to test new builds quickly and without programming EEPROM.  It doesn't need full functionality, just enough to be comfortable that buring and EEPROM should work.

> I am also had thoughts about a single rom app for both functions D and S with just a different entry point.

Yes, that makes sense.  Rather than two different entry points, I suggest setting a register to a value that is used to branch at the start of the ROM App.

> Another thought would be to load this app (once) into mid memory, but not conflict with RomLdr itself , and leave it resident, so could directly be called.

> Sort of RomLdr companion app, with non critical (to boot) functions.

Maybe.

> For app boot this would have to be stubbed out

Correct.

> Note I am physically away after Christmas, until early in new year, so wont be able to action this until then.

I am very busy as well.  This definitely can wait.

As an aside, I am very overdue to generate a stable release.  In order to do this, I need to generally freeze the repository while I work through about a month of regression testing.  Painful, but there are a lot of combinations to test.  I hope to start this process near the beginning of the year.  Work on future releases could proceed in a temporary branch.

Thanks, Wayne
